File: Ontology/Annotated.cs
using System.Collections.Generic;

namespace Ontology
{
    public abstract class Annotated: IAnnotated
    {
        private readonly IList<IAnnotation> _annotations = new List<IAnnotation>();

        protected Annotated()
        {
        }

        IList<IAnnotation> IAnnotated.Annotations => _annotations;
    }
}

File: Web/ClientApp/src/app/Ontology/Annotated.ts
import { IAnnotated } from "./IAnnotated";
import { IAnnotation } from "./IAnnotation";
import { IAnnotationProperty } from "./IAnnotationProperty";

export class Annotated implements IAnnotated
{
    readonly Annotations: IAnnotation[] = [];

    protected constructor()
    {
    }

    Annotate(
        property: IAnnotationProperty,
        value   : any): IAnnotation
    {
        let annotation = new Annotation(
            property,
            value);

        this.Annotations.push(annotation);
        return annotation;
    }
}

export class Annotation
    extends Annotated
    implements IAnnotation
{
    constructor(
        public Property: IAnnotationProperty,
        public Value   : any
        )
    {
        super();
    }
}
